### PR TITLE
add reload keyboard shortcut

### LINKED ISSUE #11686: Add a keyboard shortcut to reload the current page
We would like Ctr+R/Apple+R to reload the current page (with a preference to disable this behaviour, too). We'll need to:

1. add a new `WindowEvent` variant
1. add a handler to `Window::handle_key` that enqueues the new event type (guarded by a preference check like the one for handling escape)
1. add a new `ConstellationMsg` variant
1. add a handler for the new window event in `Compositor::handle_window_message` that sends the new message to the constellation (see `on_navigation_window_event` for a good model)
1. add a new `ConstellationControlMsg` variant that includes a `PipelineId` argument
1. add a handler for the new constellation message in `Constellation::handle_request_from_compositor` that gets the root pipeline and sends the new message to its script channel (see `Constellation::handle_key_msg` for an example), passing the root pipeline id
1. add a handler for the new message sent from the constellation to `ScriptThread::handle_msg_from_constellation` that replicates the behaviour of `handle_reload` in `devtools.rs`

There's no way to write an automated test for this as far as I know, so testing manually will be our only option here.

Code: `components/compositing/windowing.rs`, `ports/glutin/window.rs`, `components/compositing/compositor.rs`, `components/constellation/constellation.rs`, `components/script/script_thread.rs`, `components/msg/constellation_msg.rs`, `components/script_traits/lib.rs`

### PR DESCRIPTION
**Pull request by mrmiywj**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11686  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because this cannot be automated tested.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11735)

### CODE REVIEW

**Comment by KiChjang** [comment]
This fails tidy and is missing the preferences that @paulrouget specified in the issue.

**Comment by ConnorGBrewster** [comment]
From travis:
```
Checking files for tidiness...
./components/script/script_thread.rs:2054: use statement is not in alphabetical order
	expected: dom::bindings::codegen::Bindings::WindowBinding::WindowMethods
	found: dom::bindings::codegen::Bindings::LocationBinding::LocationMethods
```

**Comment by mrmiywj** [comment]
@KiChjang Sorry for that I missed this tidy

**Comment by mrmiywj** [comment]
The preference has been renamed and guarded the key shortcut.

**Comment by ConnorGBrewster** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/script_thread.rs, line 2053 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11735#-KK6vwLbVRWxvAYl-k1d:-KK6vwLbVRWxvAYl-k1e:-109787288) ([raw file](https://github.com/servo/servo/blob/90cad96efe47fbfe4268326dad269ebd06f773dd/components/script/script_thread.rs#L2053)):*
> ```Rust
> 
>     fn handle_reload(&self, pipeline_id: PipelineId) {
>         use dom::bindings::codegen::Bindings::WindowBinding::WindowMethods;
> ```

nit: could we move these `use` statements to the top of the file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11735)*

**Comment by mrmiywj** [comment]
@KiChjang I used to write `use` statements on the top of file, but I'm not sure whether it will dirty the namespace. So I used it in this clause.

**Comment by mrmiywj** [comment]
@jdm-reply

**Comment by jdm** [comment]
My apologies for the delay - I was at an event all last week. I'll review this today.

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -822,7 +822,7 @@ impl WindowMethods for Window {
             }
 
             (NONE, Key::Escape) => {
-                if let Some(true) = prefs::get_pref("shell.quit-on-escape.enabled").as_boolean() {
+                if let Some(true) = prefs::get_pref("shell.builtin-key-shortcuts.enable").as_boolean() {
```
Comment (paulrouget) [inline ports/glutin/window.rs]: s/enable/enabled

**Comment by jdm** [comment]
This looks fine; just a few small changes required to merge. Thanks for doing this work!
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/constellation/constellation.rs, line 1399 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11735#-KKsA2QnhylsuwORmrMn:-KKsA2QnhylsuwORmrMo:607448296) ([raw file](https://github.com/servo/servo/blob/90cad96efe47fbfe4268326dad269ebd06f773dd/components/constellation/constellation.rs#L1399)):*
> ```Rust
> 
>         match root_pipeline_id {
>             Some(pipeline_id) => {
> ```

This can be an `if let` instead of a `match`.

---

*[components/script/script_thread.rs, line 2055 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11735#-KKsAQ_faPc2yg_TQlvP:-KKsAQ_faPc2yg_TQlvQ:-1856374413) ([raw file](https://github.com/servo/servo/blob/90cad96efe47fbfe4268326dad269ebd06f773dd/components/script/script_thread.rs#L2055)):*
> ```Rust
>         use dom::bindings::codegen::Bindings::WindowBinding::WindowMethods;
>         use dom::bindings::codegen::Bindings::LocationBinding::LocationMethods;
>         let context = self.browsing_context.get().unwrap().find(pipeline_id).expect("There is no such context");
> ```

Let's use `self.find_child_context(pipeline_id)` instead, and just return if it can't be found.

---

*[components/script_traits/lib.rs, line 192 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11735#-KKsC_HExQfD5_84LGrD:-KKsC_HExQfD5_84LGrE:-228714125) ([raw file](https://github.com/servo/servo/blob/90cad96efe47fbfe4268326dad269ebd06f773dd/components/script_traits/lib.rs#L192)):*
> ```Rust
>     /// Report an error from a CSS parser for the given pipeline
>     ReportCSSError(PipelineId, String, usize, usize, String),
>     /// Reload the current page.
> ```

`Reload the given page`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11735#-:-KKsCw4oDID9s5GKXC7v:-616703212)*

**Comment by mrmiywj** [comment]
@jdm -reply

**Comment by paulrouget** [comment]
You also need to change `quit-on-escape` to `builtin-key-shortcuts` in:
- python/servo/package_commands.py
- python/servo/post_build_commands.py

**Comment by jdm** [comment]
@mrmiywj Just so you know, when you push new changes to a branch there's a github notification and emails that are sent out. You don't need to explicitly do the "-reply" thing each time.

**Comment by jdm** [comment]
@bors-servo: r+
Thanks @mrmiywj!

---

Reviewed 4 of 4 files at r4, 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11735#-:-KKxr11QD-zOL2mnZEon:1166668117)*

**Pull request merged**